### PR TITLE
Add validations for bmo resource name

### PR DESCRIPTION
**Pull request by hroyrh**
Add validations to check if there are any invalid character(s) in the BMO resource name.

### CODE REVIEW

**Comment by hroyrh** [comment]
fixes #855

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 1]
```diff
@@ -18,6 +19,11 @@ func (host *BareMetalHost) validateHost() []error {
 		errs = append(errs, err)
 	}
 
+	if strings.Contains(host.Name, "~") {
```
[the remark this answers is not in the document]
Comment (hroyrh) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go]: Added separate function for this.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 2]
```diff
@@ -53,3 +58,11 @@ func validateRAID(r *RAIDConfig) error {
 
 	return nil
 }
+
+func validateBMHName(bmhname string) error {
+	if strings.Contains(bmhname, "~") || bmhname == "" {
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:63]: BMH name emptiness check might belong to another PR, I'd prefer to use that way;

```
func validateBMHName(bmhname string) error {
if strings.Contains(bmhname, "~") {
		return fmt.Errorf("BareMetalHost resource name cannot be empty or contain the character `~`")
	}
return nil
}
```

Comment (hroyrh) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:63]: But wouldn't it make sense to keep all the invalid name validations together.

Comment (ardaguclu) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:63]: According to the https://github.com/kubernetes/kubernetes/blob/4753fa5dee121308581ee5cc1a790a7b87315ed8/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L124. BMH name can not be updated and it is optional.

When it is not given, it generates name https://github.com/kubernetes/kubernetes/blob/4753fa5dee121308581ee5cc1a790a7b87315ed8/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L142

IMO, there is no need to check bmh name is nil or not.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 3]
```diff
@@ -53,3 +58,11 @@ func validateRAID(r *RAIDConfig) error {
 
 	return nil
 }
+
+func validateBMHName(bmhname string) error {
+	if strings.Contains(bmhname, "~") {
```
[the remark this answers is not in the document]
Comment (zaneb) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:63]: We can't import from there into the api module.

**Review by zaneb — COMMENTED** [review]
This looks fine, but given that the two name validation tasks are:

* Resource name that is invalid in ironic (looks like a UUID or contains chars other than `[A-Za-z0-9~._-]`)
* Resource name that contains a ~ (used in ironic Node name to separate name + namespace)

it would make more sense to me if we implemented both at the same time, since the solution will clearly be a regex and not `strings.Contains()`.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 4]
```diff
@@ -60,8 +61,15 @@ func validateRAID(r *RAIDConfig) error {
...
-		return fmt.Errorf("BareMetalHost resource name cannot contain the character `~`")
+
+	invalidname, _ := regexp.MatchString(`[^A-Za-z0-9\.\-\_]`, bmhname)
+	if invalidname {
+		return fmt.Errorf("BareMetalHost resource name cannot contain characters other than [A-Za-z0-9._-]")
+	}
+
+	_, err := uuid.Parse(bmhname)
+	if err == nil {
+		return fmt.Errorf("BareMetalHost resource name cannot be a UUID")
```
[the remark this answers is not in the document]
Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:72]: Ironic prevents node names from looking like UUIDs to avoid ambiguity. I think BMH names become Ironic names (modulo a namespace).

**Comment by andfasano** [comment]
Given that the validation rules are fine, my biggest concern here is that they are not applied in the operator (ie, in the create event filter), thus leading to accept invalid BMH when webhook is not enabled / not available. Thoughts?

**Comment by ardaguclu** [comment]
> Given that the validation rules are fine, my biggest concern here is that they are not applied in the operator (ie, in the create event filter), thus leading to accept invalid BMH when webhook is not enabled / not available. Thoughts?

@andfasano baremetal operator webhook failure policy(both upstream and downstream) is set to fail. That means if webhook has problems, operator can not accept any requests. Or you mean somewhere else that did not come to my mind?

**Comment by andfasano** [comment]
> > Given that the validation rules are fine, my biggest concern here is that they are not applied in the operator (ie, in the create event filter), thus leading to accept invalid BMH when webhook is not enabled / not available. Thoughts?
> 
> @andfasano baremetal operator webhook failure policy(both upstream and downstream) is set to fail. That means if webhook has problems, operator can not accept any requests. Or you mean somewhere else that did not come to my mind?

I'm referring to the _"Validating a resource in the controller"_ chapter of the ["Kubernetes operator best practices" blog article](https://cloud.redhat.com/blog/kubernetes-operators-best-practices) where essentially the recommendation it's to organize the code to share the same validation rule between the controller and the webhook, so that in those cases where the webhook is not available (maybe because not enabled or not yet started due some racy conditions) the check will not be missed. In addition, I'd say that sharing the code in such case will ensure that the validation logic will not diverge between the controller and the webhook.

**Comment by ardaguclu** [comment]
Maybe we can discuss this again after bmc module created.
/lgtm
/test all

**Comment by andfasano** [comment]
> Maybe we can discuss this again after bmc module created. /lgtm /test all

Despite the technical difficulties, my point was to decide a common and shared approach for this specific point at design level, given that also in other BMO controllers we are currently going to add webhooks.
